[Backport-2.2] Code generation improvement for php 7.1

swnsma opened the pull request:
Code generation improvements.
### Description
Improve code generation for proxies and interceptors.
Fix not supported nullable return types and parameters in interceptors, variadics and return types like 'self'.
Make changes compatible with PHP 7.0.
Rework of #16337

### Fixed Issues
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento/magento2#15505: Interceptor class methods do not support nullable return types
2. MAGETWO-89899: Proxy/Interceptor generator does not understand PHP 7.1 syntax
3. MAGETWO-95415: Can't generate Plugin if object method returns self

### Manual testing scenarios
Create class with usage:
1. variadics
2. nullable type return/parameter
3. return type 'self'
Check in scope of PHP7.0 and PHP7.1

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Code generate: support variadic parameter

Commit:
Fix integration test errors

Commit:
MAGETWO-89899: Proxy generator does not understand PHP 7.1 syntax

Commit:
MAGETWO-89899: Proxy generator does not understand PHP 7.1 syntax

Commit:
MAGETWO-89899: Proxy/Interceptor generator does not understand PHP 7.1 syntax

Commit:
MAGETWO-89899: Proxy/Interceptor generator does not understand PHP 7.1 syntax

Commit:
MAGETWO-89899: Proxy/Interceptor generator does not understand PHP 7.1 syntax

Commit:
MAGETWO-95415: Can't generate Plugin if object method returns self

Commit:
magento/magento2##15505: Interceptor class methods do not support nullable return types.
Backport variadic, self return types compatibility for proxy/interceptor code generations.
Backport nullable type hints for PHP7.1

swnsma committed:
Merge branch '2.2-develop' into Code-Generation-Improvement-For-PHP-7.1

orlangur commented on lib/internal/Magento/Framework/Interception/Code/Generator/Interceptor.php:
```diff
@@ -6,6 +6,11 @@
 
+/**
+ * Class Interceptor
+ ˚*
```
Some junk.

orlangur commented on dev/tests/integration/testsuite/Magento/Framework/Code/GeneratorTest.php:
```diff
@@ -172,6 +173,74 @@ public function testGenerateClassExtensionAttributesInterfaceFactoryWithNamespac
+    public function testGenerateClassProxyWithNamespace71()
+    {
+        if (!$this->shouldRun71()) {
+            $this->assertTrue(true);
```
`markTestSkipped` must be used in case of lower PHP version instead, please fix all occurrences.

orlangur reviewed (changes requested):
Great @swnsma, almost perfect! Thanks for combining so many improvements in a single backport.

Commit:
magento/magento2##15505: Interceptor class methods do not support nullable return types.
- removed junk;
- use 'required' PHP_Unit directive to skip tests, which require PHP 7.1.

Commit:
Merge branch 'Code-Generation-Improvement-For-PHP-7.1' of github.com:swnsma/magento2-1 into Code-Generation-Improvement-For-PHP-7.1

swnsma commented:
Hi @orlangur,
Thank you for your review.
Requested changed have been applied.

orlangur reviewed (approved):
Wow, man, I didn't even know of this feature :O 👍 

https://phpunit.readthedocs.io/en/7.3/incomplete-and-skipped-tests.html#incomplete-and-skipped-tests-requires-tables-api looks like a lot of manual checks in our tests may be refactored.

magento-engcom-team commented:
Hi @orlangur, thank you for the review.
[ENGCOM-3568](https://jira.corp.magento.com/browse/ENGCOM-3568) has been created to process this Pull Request

magento-engcom-team commented:
Hi @swnsma. Thank you for your contribution. 
We will aim to release these changes as part of 2.2.8. 
Please check the release notes for final confirmation.

Please, consider to port this solution to 2.3 release line.
You may use [Porting tool](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#porting-tool) to port commits automatically.
